### ISSUE TITLE
Crash on android

### ISSUE THREAD

**Issue by hugleo** [koreader/koreader#12332]
On the last or previous nightly.

Onyx Android 11

08-11 10:11:53.539  6478  6478 I MainActivity: loading libluajit
08-11 10:11:53.561  6478  7175 V NativeGlue: User FIFO already exists at `/data/user/0/org.koreader.launcher/files/alooper.fifo`
08-11 10:11:53.561  6478  7175 V NativeThread: waiting for activity
08-11 10:11:53.576  6478  6478 V MainActivity: native orientation: portrait
08-11 10:11:53.585  6478  6478 V NativeGlue: Start: 0xb4000079673fb9b0
08-11 10:11:53.586  6478  6478 V NativeGlue: Resume: 0xb4000079673fb9b0
08-11 10:11:53.594  6478  6478 V NativeGlue: InputQueueCreated: 0xb4000079673fb9b0 -- 0xb400007977435290
08-11 10:11:53.632  6478  6478 V NativeGlue: NativeWindowCreated: 0xb4000079673fb9b0 -- 0xb400007a27413d00
08-11 10:11:53.632  6478  7175 V NativeThread: Activity window ready.
08-11 10:11:53.643  6478  6478 V NativeGlue: NativeWindowResized: 0xb4000079673fb9b0 -- 0xb400007a27413d00
08-11 10:11:53.655  6478  6478 V NativeGlue: NativeWindowRedrawNeeded: 0xb4000079673fb9b0 -- 0xb400007a27413d00
08-11 10:11:53.761  6478  6478 V NativeGlue: WindowFocusChanged: 0xb4000079673fb9b0 -- 1
08-11 10:11:53.762  6478  7175 V NativeThread: Activity gained focus.
08-11 10:11:53.766  6478  7175 V NativeThread: Launching LuaJIT assets
08-11 10:11:53.795  6478  7175 E NativeThread: Failed to run script: [string "android.lua"]:1300: table overflow
08-11 10:11:53.824  6478  6478 V NativeGlue: NativeWindowResized: 0xb4000079673fb9b0 -- 0xb400007a27413d00
08-11 10:11:53.873  6478  6478 V NativeGlue: NativeWindowRedrawNeeded: 0xb4000079673fb9b0 -- 0xb400007a27413d00
08-11 10:11:53.952  6478  6478 V NativeGlue: Pause: 0xb4000079673fb9b0
08-11 10:11:53.970  6478  7177 F libc    : FORTIFY: pthread_mutex_lock called on a destroyed mutex (0x7b99a0f0b0)
08-11 10:11:53.970  6478  7178 F libc    : FORTIFY: pthread_mutex_lock called on a destroyed mutex (0x7b99a0f0b0)
08-11 10:11:53.972  6478  7178 F libc    : Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 7178 (hwuiTask1), pid 6478 (reader.launcher)
08-11 10:11:54.042  1374  1412 W ActivityManager: setHasOverlayUi called on unknown pid: 6478

--------- beginning of main
08-11 10:22:38.214 I/MainActivity( 7193): loading libluajit
08-11 10:22:38.215 W/linker  ( 7193): Warning: "/data/app/~~Ui7QGxy6WKJ65eH4_cG2aw==/org.koreader.launcher-mAkrauake61IsiCNRb7gnQ==/lib/arm64/libluajit.so" unused DT entry: DT_RPATH (type 0xf arg 0xbc2) (ignoring)
08-11 10:22:38.238 V/NativeGlue( 7193): User FIFO already exists at `/data/user/0/org.koreader.launcher/files/alooper.fifo`
08-11 10:22:38.238 V/NativeThread( 7193): waiting for activity
08-11 10:22:38.238 I/Assets  ( 7193): loading lib7z
08-11 10:22:38.242 I/EPD     ( 7193): Using Onyx/Qualcomm driver
08-11 10:22:38.243 I/Lights  ( 7193): Using Onyx Adb driver
08-11 10:22:38.248 V/Surface ( 7193): Using SurfaceView implementation
08-11 10:22:38.248 V/MainActivity( 7193): native orientation: portrait
08-11 10:22:38.248 V/EventReceiver( 7193): Filtering 3 events: 
08-11 10:22:38.248 V/EventReceiver( 7193): android.intent.action.ACTION_POWER_CONNECTED
08-11 10:22:38.248 V/EventReceiver( 7193): android.intent.action.ACTION_POWER_DISCONNECTED
08-11 10:22:38.248 V/EventReceiver( 7193): android.intent.action.DOWNLOAD_COMPLETE
08-11 10:22:38.252 V/NativeGlue( 7193): Start: 0xb400007967411170
08-11 10:22:38.253 V/NativeGlue( 7193): Resume: 0xb400007967411170
08-11 10:22:38.254 V/Timeout ( 7193): onResume: updating system timeout: 10
08-11 10:22:38.254 V/Timeout ( 7193): onResume
08-11 10:22:38.259 V/NativeGlue( 7193): InputQueueCreated: 0xb400007967411170 -- 0xb40000797742b6d0
--------- beginning of system
08-11 10:22:38.261 V/ActivityThread( 7193): reportTopResumedActivityChanged - > onActivityTopResumed
08-11 10:22:38.267 D/Surface ( 7193): onAttachedToWindow()
08-11 10:22:38.288 V/NativeGlue( 7193): NativeWindowCreated: 0xb400007967411170 -- 0xb400007a27413d00
08-11 10:22:38.288 V/NativeThread( 7193): Activity window ready.
08-11 10:22:38.309 V/Surface ( 7193): surface changed {
08-11 10:22:38.309 V/Surface ( 7193):   width:  1860
08-11 10:22:38.309 V/Surface ( 7193):   height: 2354
08-11 10:22:38.309 V/Surface ( 7193):  format: TRANSPARENT
08-11 10:22:38.309 V/Surface ( 7193): }
08-11 10:22:38.310 V/NativeGlue( 7193): NativeWindowResized: 0xb400007967411170 -- 0xb400007a27413d00
08-11 10:22:38.324 V/NativeGlue( 7193): NativeWindowRedrawNeeded: 0xb400007967411170 -- 0xb400007a27413d00
08-11 10:22:38.437 V/NativeGlue( 7193): WindowFocusChanged: 0xb400007967411170 -- 1
08-11 10:22:38.437 V/NativeThread( 7193): Activity gained focus.
08-11 10:22:38.443 V/NativeThread( 7193): Launching LuaJIT assets
08-11 10:22:38.459 E/NativeThread( 7193): Failed to run script: [string "android.lua"]:1300: table overflow
08-11 10:22:38.498 V/Surface ( 7193): surface changed {
08-11 10:22:38.498 V/Surface ( 7193):   width:  1860
08-11 10:22:38.498 V/Surface ( 7193):   height: 2480
08-11 10:22:38.498 V/Surface ( 7193):  format: TRANSPARENT
08-11 10:22:38.498 V/Surface ( 7193): }
08-11 10:22:38.501 V/NativeGlue( 7193): NativeWindowResized: 0xb400007967411170 -- 0xb400007a27413d00

@benoit-pierre Could be one of yours patches?

**Comment by hugleo** [comment]
Oh,

Build the nightly myself and is not crashing.

Can be something wrong with the OTA version.

**Comment by benoit-pierre** [comment]
It starts fine here (the ARM nightly), but the code is assuming it's a new install on every startup… Even though `/data/user/0/org.koreader.launcher.debug/files/git-rev` exists.

**Comment by benoit-pierre** [comment]
`module/version.txt` is missing…

**Comment by benoit-pierre** [comment]
As for the crash, I need more detailed logs. Did you uninstall the previous version before updating?

**Comment by benoit-pierre** [comment]
I downgraded to v2024.07, then updated to v2024.07-79-gbd4a651fb without uninstalling the previous version, and it works fine (except for that always slow startup, fixed in #12333).

**Comment by hugleo** [comment]
> As for the crash, I need more detailed logs. Did you uninstall the previous version before updating?

No,

Don't remember the exactly version but my previous nightly shouldn't be more than two weeks older.

As for my manually build yes (run a debug version).

**Comment by hugleo** [comment]
I will try to downgrade

**Comment by hugleo** [comment]
Weird,

After install https://build.koreader.rocks/download/nightly/v2024.07-79-gbd4a651fb_2024-08-10/koreader-android-arm64-v2024.07-79-gbd4a651fb_2024-08-10.apk I can't downgrade to any version, even stable (similar behavior as if was a different key signed apk).
But android let install koreader-android-arm64-v2024.07-79-gbd4a651fb_2024-08-10.apk again

**Comment by hugleo** [comment]
Tried on my mobile phone now.
Same crash happened.
Before version: 2024.03.1-36
Updated via OTA.

**Comment by hugleo** [comment]
I I uninstall and install directly koreader-android-arm64-v2024.07-79-gbd4a651fb_2024-08-10.apk no crash.

I I uninstall and install koreader-android-arm64-v2024.07.apk for example and install koreader-android-arm64-v2024.07-79-gbd4a651fb_2024-08-10.apk will crash.

Tried with the previous nightly koreader-android-arm64-v2024.07-76-gdc3b63837_2024-08-09.apk  and no crash.

So the problem is with koreader-android-arm64-v2024.07-79-gbd4a651fb_2024-08-10.apk

**Comment by benoit-pierre** [comment]
You still have not provided more detailed logs. If you have a dev environment:
- install the nightly
- clear logcat buffer: `adb logcat -c`
- use `./kodev log -d android | tee crash.log`
- and start koreader

**Comment by hugleo** [comment]
follow

[crash.log](https://github.com/user-attachments/files/16574100/crash.log)

**Comment by benoit-pierre** [comment]
Looks like this is the problem:
```
08-11 12:04:31.050 11873 11903 E NativeThread: Failed to run script: [string "android.lua"]:1300: table overflow
```

**Comment by ilyats** [comment]
Crash happened to me on Onyx Page too. After uninstalling the app, installing the previous build and updating back to latest build disappeared.

**Comment by benoit-pierre** [comment]
I wonder if this is `-ffast-math` being problematic again.

**Comment by benoit-pierre** [comment]
Can you reproduce if you build a release version? (To sign the APK with debug keys you can use [uber-apk-signer](https://github.com/patrickfav/uber-apk-signer): `uber-apk-signer --overwrite -a android.apk`).

**Comment by benoit-pierre** [comment]
I myself can't reproduce the issue on my Nexus 7 (Android 11).

**Comment by hugleo** [comment]
> Can you reproduce if you build a release version? (To sign the APK with debug keys you can use [uber-apk-signer](https://github.com/patrickfav/uber-apk-signer): `uber-apk-signer --overwrite -a android.apk`).

I'll do that

**Comment by benoit-pierre** [comment]
And here is a [release build](https://ufile.io/khho8lvp) without `-ffast-math` (signed with debug keys).

**Comment by hugleo** [comment]
For test I need another version with the same debug keys, for example tag https://github.com/koreader/koreader/releases/tag/v2024.07

**Comment by benoit-pierre** [comment]
Why?

**Comment by hugleo** [comment]
Because the problem only happens If I update from some version to the last. If I install the last version from a clean uninstall the crash will not occur.

**Comment by benoit-pierre** [comment]
Here's a [version of v2024.07](https://ufile.io/0xb7dw5a) re-signed with the same debug keys.

**Comment by hugleo** [comment]
No crash with these versions

**Comment by benoit-pierre** [comment]
OK, let me rebuild with `-ffast-math`: if you still can't  reproduce, will chalk it up to #12333.

Here: https://ufile.io/s08ec3l7.

**Comment by hugleo** [comment]
btw my local release build crashed

**Comment by benoit-pierre** [comment]
OK, you can try:
- updating to de2a9d54f1 (which contains #12333) if your version was older
- re-building without `-ffast-math` by patching `base/Makefile.defs`:
```diff
--- a/Makefile.defs
+++ b/Makefile.defs
@@ -359,11 +359,11 @@ endif
 
 # Base CFLAGS, without arch. We'll need it for luajit, because its
 # Makefiles do some tricky stuff to differentiate HOST/TARGET
-BASE_CFLAGS:=-O2 -ffast-math -pipe -fomit-frame-pointer
+BASE_CFLAGS:=-O2 --pipe -fomit-frame-pointer
 # For ricers.
 #BASE_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb
 # Used to enforce aggressive optimizations in some components
-VECTO_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb -ftree-vectorize -funroll-loops
+VECTO_CFLAGS:=-O3 -pipe -fomit-frame-pointer -frename-registers -fweb -ftree-vectorize -funroll-loops
 # And for an aggressive LTO build
 #BASE_CFLAGS:=$(VECTO_CFLAGS) -flto$(if $(PARALLEL_JOBS),=$(PARALLEL_JOBS)) -fuse-linker-plugin
 # Use this for debugging:
```

**Comment by hugleo** [comment]
> OK, let me rebuild with `-ffast-math`: if you still can't reproduce, will chalk it up to #12333.
> 
> Here: https://ufile.io/s08ec3l7.

No crash here

**Comment by benoit-pierre** [comment]
OK, so let's assume #12333 was to blame, and `-ffast-math` get to live another day.

**Comment by hugleo** [comment]
> OK, you can try:
> 
>     * updating to [de2a9d5](https://github.com/koreader/koreader/commit/de2a9d54f1de55a28d5f667a4857af0fc1febc7e) (which contains [fix missing android APK asset #12333](https://github.com/koreader/koreader/pull/12333)) if your version was older
> 
>     * re-building without `-ffast-math` by patching `base/Makefile.defs`:
> 
> 
> ```diff
> --- a/Makefile.defs
> +++ b/Makefile.defs
> @@ -359,11 +359,11 @@ endif
>  
>  # Base CFLAGS, without arch. We'll need it for luajit, because its
>  # Makefiles do some tricky stuff to differentiate HOST/TARGET
> -BASE_CFLAGS:=-O2 -ffast-math -pipe -fomit-frame-pointer
> +BASE_CFLAGS:=-O2 --pipe -fomit-frame-pointer
>  # For ricers.
>  #BASE_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb
>  # Used to enforce aggressive optimizations in some components
> -VECTO_CFLAGS:=-O3 -ffast-math -pipe -fomit-frame-pointer -frename-registers -fweb -ftree-vectorize -funroll-loops
> +VECTO_CFLAGS:=-O3 -pipe -fomit-frame-pointer -frename-registers -fweb -ftree-vectorize -funroll-loops
>  # And for an aggressive LTO build
>  #BASE_CFLAGS:=$(VECTO_CFLAGS) -flto$(if $(PARALLEL_JOBS),=$(PARALLEL_JOBS)) -fuse-linker-plugin
>  # Use this for debugging:
> ```

I'll try this and let you know

**Comment by hugleo** [comment]
Edited my previous message

**Comment by benoit-pierre** [comment]
Is `platform/android/luajit-launcher` clean? With https://github.com/koreader/koreader/commit/bd4a651fbf8e95fd3f5428251a2ff975db83a165, the only thing that should be created after a build is an empty `.gradle/kotlin/errors` directory.

**Comment by benoit-pierre** [comment]
I can reproduce:
```
08-11 14:21:02.763 28515 28543 I Assets  : Skip installation for revision v2024.07-82-gde2a9d54f_2024-08-11
```
Because the same version was reinstalled, assets are not installed again, which mean the `libs` symlink is not updated…

**Comment by benoit-pierre** [comment]
Neither is the `sdcv` symlink, BTW. Which is also an issue previous to https://github.com/koreader/koreader/commit/bd4a651fbf8e95fd3f5428251a2ff975db83a165: install v2024.07 2 times, and dictionary lookups stop working!

**Comment by hugleo** [comment]
An observation:

The problem occurs only for arm64 builds.

I can't reproduce it using an arm build either. That's is because @benoit-pierre apk arm32 didn't crash.

**Comment by benoit-pierre** [comment]
Yes, I can reproduce with an arm64 build.

**Comment by benoit-pierre** [comment]
Closed in #12360.
